### PR TITLE
Position models

### LINKED ISSUE #7: Implement position model to allow dense detector arrays
**Scope**
Primarily muondetector-daemon, for accessibility and monitoring also muondetector-gui

**Description**
Currently, the geographic coordinates of a detector station are determined by sending the mean of the coordinates from continuous GNSS fixes over a one-minute interval (MUONPI_LOG_INTERVAL_MINUTES in config.h) through log messages to the server. Therefore, these short-time fixes may wander around over several meters or even more depending on the GNSS reception. This scheme can not be used for the Time-of-Flight (ToF) reconstruction of showers in dense close-vicinity constellations of detectors. A fixed geometrical arrangement with relative positions to one precise reference coordinate is required for this purpose.

**Proposition**
In order to accomodate for the dense-detector arrangement with the (hardware-)unmodified MuonPi detectors, the geographic coordinates reported by each station could be altered. For this end, the following proposition is made:

Introduce a GNSS fix mode which may assume several states: 
- auto mode which determines and reports the position as it is currently done (default).
- static mode, where a fixed set of geo coordinates is provided (eg. from config file and/or the GUI). Then this set of coordinates is logged and published instead of the GNSS fixes
- lock-in mode, where the daemon accumulates all fixes over a specified measurement interval (LOCK_IN_MEASUREMENT_INTERVAL probably to be set in config.h) on the order of typically a day. In the measurement phase, the fixes are only accepted and stored to the accumulation buffer, if a limiting fix precision is kept (e.g. LOCK_IN_FIX_LIMIT_GEO,  LOCK_IN_FIX_LIMIT_TIME, LOCK_IN_FIX_LIMIT_DOP in config.h or something of that kind). In the end of the lock-in phase, the mean (median?) of all coordinates is assumed to be the most precise determination of the station's position and overrides the static geo coordinates. Those will from then on be reported like in static mode.

With the static or lock-in mode activated, each station may fix it's position more accurate than in free-floating auto mode. For dense detector arrays, however the static mode would seem more preferable. Setting the coordinates for each station could then be aleviated by a small add-on in the GUI, where the station's coordinates can be calculated and set by specifying a reference coordinate and a relative geometric offset.

Any more ideas or proposals on this?

### PR DESCRIPTION
**Pull request by hangeza**
PR to merge all the functionality of position model handling into dev

### CODE REVIEW

File: daemon/include/daemon.h [hunk 1]
```diff
@@ -257,16 +269,21 @@ private slots:
     QTimer rateBufferReminder;
     QTimer oledUpdateTimer;
     QList<quint64> andCounts, xorCounts;
-    UbxDopStruct currentDOP;
-    Property nrSats, nrVisibleSats, fixStatus;
+
+    Property<size_t> nrSats {};
+    Property<size_t> nrVisibleSats {};
+    Property<UbxDopStruct> currentDOP {};
```
Comment (dtreffenstaedt) [inline daemon/include/daemon.h]: Fix naming of these class members.

File: daemon/include/daemon.h [hunk 2]
```diff
@@ -257,16 +269,21 @@ private slots:
...
+
     QVector<QTcpSocket*> peerList;
     QList<float> adcSamplesBuffer;
     ADC_SAMPLING_MODE adcSamplingMode { ADC_SAMPLING_MODE::PEAK };
     int currentAdcSampleIndex { -1 };
     QTimer samplingTimer;
     QTimer parameterMonitorTimer;
     QTimer rateScanTimer;
-    QMap<QString, Property> propertyMap;
+    //    QMap<QString, Property> propertyMap;
```
Comment (dtreffenstaedt) [inline daemon/include/daemon.h]: Do not uncomment code. Remove instead.

File: daemon/include/utility/kalman_gnss_filter.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    /// <param name="alt_measurement">new measurement of altitude</param>
+    /// <param name="accuracy">measurement of 1 standard deviation error in metres</param>
+    /// <returns>new state</returns>
+    void process(double lat_measurement, double lng_measurement, double alt_measurement, double accuracy);
+    [[nodiscard]] auto get_timestamp() const -> std::chrono::time_point<std::chrono::steady_clock> { return m_timestamp; }
+    [[nodiscard]] auto get_latitude() const -> double { return m_lat; }
+    [[nodiscard]] auto get_longitude() const -> double { return m_lng; }
+    [[nodiscard]] auto get_altitude() const -> double { return m_alt; }
+    [[nodiscard]] auto get_accuracy() const -> double { return std::sqrt(m_variance); }
+    [[nodiscard]] auto valid() const -> bool { return (m_variance >= 0.); }
```
Comment (dtreffenstaedt) [inline daemon/include/utility/kalman_gnss_filter.h]: Move implementations to the source file.

File: daemon/src/daemon.cpp [hunk 4]
```diff
@@ -1302,41 +1339,207 @@ void Daemon::receivedCalibItems(const std::vector<CalibStruct>& newCalibs)
...
+}
+
+void Daemon::sendPositionModel(const PositionModeConfig& pos)
+{
+    TcpMessage tcpMessage(TCP_MSG_KEY::MSG_POSITION_MODEL);
+    (*tcpMessage.dStream) << pos;
+    emit sendTcpMessage(tcpMessage);
+}
+
+void Daemon::onGpsPropertyUpdatedGeodeticPos(const GnssPosStruct& pos)
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp:1356]: This function is way too long and complex. Split it into smaller parts to make it readable and maintainable.

File: daemon/src/daemon.cpp [hunk 5]
```diff
@@ -1917,12 +2120,13 @@ void Daemon::onGpsPropertyUpdatedGnss(const std::vector<GnssSatellite>& sats,
...
-    nrVisibleSats = QVariant { static_cast<qulonglong>(visibleSats.size()) };
-
+    nrSats = Property<size_t>("nrSats", N);
+    nrVisibleSats = Property<size_t>("visSats", visibleSats.size());
+    ;
+    /*
     propertyMap["nrSats"] = Property("nrSats", N);
     propertyMap["visSats"] = Property("visSats", visibleSats.size());
-
+*/
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 6]
```diff
@@ -1932,9 +2136,10 @@ void Daemon::onGpsPropertyUpdatedGnss(const std::vector<GnssSatellite>& sats,
                 maxCnr = sat.Cnr;
         }
     }
+    /*
     propertyMap["usedSats"] = Property("usedSats", usedSats);
     propertyMap["maxCNR"] = Property("maxCNR", maxCnr);
-
+*/
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 7]
```diff
@@ -2034,8 +2239,9 @@ void Daemon::gpsPropertyUpdatedUint8(uint8_t data, std::chrono::duration<double>
         delete tcpMessage;
         emit logParameter(LogParameter("fixStatus", QString::number(data), LogParameter::LOG_LATEST));
         emit logParameter(LogParameter("fixStatusString", QString::fromLocal8Bit(Gnss::FixType::name[data]), LogParameter::LOG_LATEST));
-        fixStatus = QVariant(data);
-        propertyMap["fixStatus"] = Property("fixStatus", QString::fromLocal8Bit(Gnss::FixType::name[data]));
+        //        fixStatus = Property<std::string>("fixStatus", std::string(Gnss::FixType::name[data]));
+        m_fix_status = Property<Gnss::FixType>("fixStatus", { data });
+        //propertyMap["fixStatus"] = Property("fixStatus", QString::fromLocal8Bit(Gnss::FixType::name[data]));
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 8]
```diff
@@ -2085,7 +2292,7 @@ void Daemon::gpsPropertyUpdatedUint32(uint32_t data, chrono::duration<double> up
             std::cout << std::chrono::system_clock::now()
                     - std::chrono::duration_cast<std::chrono::microseconds>(updateAge)
                       << "rising edge counter: " << data << std::endl;
-        propertyMap["events"] = Property("events", (quint16)data);
+        //propertyMap["events"] = Property("events", (quint16)data);
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 9]
```diff
@@ -2106,15 +2313,15 @@ void Daemon::gpsPropertyUpdatedInt32(int32_t data, std::chrono::duration<double>
                     - std::chrono::duration_cast<std::chrono::microseconds>(updateAge)
                       << "clock drift: " << data << " ns/s" << std::endl;
         logParameter(LogParameter("clockDrift", QString::number(data) + " ns/s", LogParameter::LOG_AVERAGE));
-        propertyMap["clkDrift"] = Property("clkDrift", (qint32)data);
+        //propertyMap["clkDrift"] = Property("clkDrift", (qint32)data);
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 10]
```diff
@@ -2106,15 +2313,15 @@ void Daemon::gpsPropertyUpdatedInt32(int32_t data, std::chrono::duration<double>
...
         break;
     case 'b':
         if (verbose > 3)
             std::cout << std::chrono::system_clock::now()
                     - std::chrono::duration_cast<std::chrono::microseconds>(updateAge)
                       << "clock bias: " << data << " ns" << std::endl;
         emit logParameter(LogParameter("clockBias", QString::number(data) + " ns", LogParameter::LOG_AVERAGE));
-        propertyMap["clkBias"] = Property("clkBias", (qint32)data);
+        //propertyMap["clkBias"] = Property("clkBias", (qint32)data);
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 11]
```diff
@@ -2320,6 +2531,28 @@ void Daemon::aquireMonitoringParameters()
             logParameter(LogParameter("vadc4", QString::number(v2) + " V", LogParameter::LOG_AVERAGE));
         }
     }
+
+    switch (config.position_mode_config.mode) {
+    case PositionModeConfig::Mode::Static:
+        if (config.position_mode_config.static_position.valid()) {
+            //sendGeodeticPos(config.position_mode_config.static_position.getPosStruct());
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: Do not comment code. Remove instead.

File: daemon/src/daemon.cpp [hunk 12]
```diff
@@ -2320,6 +2531,28 @@ void Daemon::aquireMonitoringParameters()
             logParameter(LogParameter("vadc4", QString::number(v2) + " V", LogParameter::LOG_AVERAGE));
         }
     }
+
+    switch (config.position_mode_config.mode) {
```
Comment (dtreffenstaedt) [inline daemon/src/daemon.cpp]: This does not need to be a switch statement. It only has one case.

File: daemon/src/utility/kalman_gnss_filter.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+void KalmanGnssFilter::reset()
+{
+    m_lat = m_lng = m_alt = 0.;
+    m_variance = -1.;
+}
+
+void KalmanGnssFilter::process(double lat_measurement, double lng_measurement, double alt_measurement, double accuracy)
+{
+    double local_accuracy { accuracy };
+    if (accuracy < c_min_accuracy)
```
Comment (dtreffenstaedt) [inline daemon/src/utility/kalman_gnss_filter.cpp]: Missing curly braces. Do not add statements without them.

File: daemon/src/utility/kalman_gnss_filter.cpp [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    m_lat = lat;
+    m_lng = lng;
+    m_alt = alt;
+    m_variance = accuracy * accuracy;
+    m_timestamp = std::chrono::steady_clock::now();
+}
+
+void KalmanGnssFilter::reset()
+{
+    m_lat = m_lng = m_alt = 0.;
```
Comment (dtreffenstaedt) [inline daemon/src/utility/kalman_gnss_filter.cpp:54]: No.

File: gui/qml/CustomMap.qml [hunk 15]
```diff
@@ -14,49 +14,71 @@ CustomMapForm {
...
+    }
+    function setCoordinates(lon,lat,alt,hor_err,vert_err,tot_err)
+    {
+        map.onCoordsReceived(lon,lat,alt,hor_err,vert_err,tot_err)
     }
     function setEnabled(enabled)
     {
         map.onEnable(enabled)
     }
+    signal coordinateSignal(double lat, double lon)
```
Comment (dtreffenstaedt) [inline gui/qml/CustomMap.qml]: Keep the order of parameters consistent across the codebase.
Ideally use a struct instead.

File: gui/src/map.cpp [hunk 16]
```diff
@@ -5,6 +5,7 @@
 #include <QSsl>
 #include <cmath>
 #include <map.h>
+#include <muondetector_structs.h>
```
Comment (dtreffenstaedt) [inline gui/src/map.cpp]: Use double quotes instead of angle brackets for local includes.

File: gui/src/map.cpp [hunk 17]
```diff
@@ -29,16 +39,41 @@ Map::~Map()
...
+}
+
+void Map::onGeodeticPosReceived(const GnssPosStruct& pos)
 {
     if (mapComponent == nullptr) {
         return;
     }
-    double pos_accuracy { std::sqrt(pos.hAcc * pos.hAcc + pos.vAcc * pos.vAcc) };
-    QMetaObject::invokeMethod(mapComponent, "setCircle",
+    double pos_error { std::sqrt(pos.hAcc * pos.hAcc + pos.vAcc * pos.vAcc) };
```
Comment (dtreffenstaedt) [inline gui/src/map.cpp]: Make this const.

File: library/include/muondetector_structs.h [hunk 18]
```diff
@@ -191,5 +244,129 @@ class Property {
...
+
+    std::string name { };
+    std::string unit { };
+
+private:
+    std::any m_value {};
+    bool m_updated { false };
+    std::chrono::time_point<std::chrono::steady_clock> m_update_time { };
+};
+*/
```
Comment (dtreffenstaedt) [inline library/include/muondetector_structs.h]: Do not comment code. Remove instead.

**Review by dtreffenstaedt — CHANGES_REQUESTED** [review]
Fix all comments
Use clang-tidy
Use clang-format

**Comment by hangeza** [comment]
Thanks for the review. Will try to fix all marked items. However, clang-format was applied regularly, so the PR should not have issues with this.

File: daemon/src/utility/kalman_gnss_filter.cpp (hunk 14 above)
Comment (hangeza) [reply daemon/src/utility/kalman_gnss_filter.cpp:54]: I like the readability of such in-line assignments. Is there a good reason, not to do this?

**Comment by hangeza** [comment]
I was wrongly merging position-models into dev. Intention was just to merge dev into position-models to sync both branches but fucked it up. Now the branch is merged and i have no idea how to revert it. Advice appreciated...
I think, however, that all but one item marked in the review of @dtreffenstaedt were settled, so merging shouldn't have done harm.

**Comment by dtreffenstaedt** [comment]
All items should have been fixed.

**Comment by hangeza** [comment]
Well, all but this [one](https://github.com/MuonPi/muondetector/pull/83#discussion_r927516321), which i don't understand that it is an issue. Could you elaborate on this?

**Comment by hangeza** [comment]
As i understand now, the PR was closed but the branch "position-models" is NOT merged into dev. How should we proceed? Is it possible to reopen the PR?

**Pull request merged**